doc/reference/index.rst

```diff
@@ -23,3 +23,214 @@ API Reference
+  thread to sleep
+* :ref:`api_term_no-wait` if a parameter to the function can prevent the
+  invoking thread from trying to sleep
+* :ref:`api_term_isr-ok` if the function can always be safely called from non-thread
+  (interrupt) context even if it may return an error in that case
+* :ref:`api_term_pre-kernel-ok` if the function can be safely called
+  before the kernel has been fully initialized, even if it may return an
+  error in that case
+* :ref:`api_term_async` if the function may return before the operation
+  it initializes is complete (i.e. function return and operation
+  completion are asynchronous)
+* :ref:`api_term_supervisor` if the calling thread must have supervisor
+  privileges to execute the function
+
+Details on the behavioral impact of each attribute are in the following
+sections.  Attributes for some common kernel API are below:
+
++-------------------------------+------------+-------+---------+--------+-------+-----------+
+|name                           | reschedule | sleep | no-wait | isr-ok | async | supervisor|
++===============================+============+=======+=========+========+=======+===========+
+|:c:func:`k_busy_wait`          | no         | no*   | no*     | yes*   | no    | no        |
++-------------------------------+------------+-------+---------+--------+-------+-----------+
+|:c:func:`k_sem_give`           | yes        | no    | no*     | yes*   | no    | no        |
++-------------------------------+------------+-------+---------+--------+-------+-----------+
+|:c:func:`k_sleep`              | yes*       | yes   | no      | no     | no    | no        |
++-------------------------------+------------+-------+---------+--------+-------+-----------+
+|:c:func:`k_sem_take`           | yes*       | yes   | yes     | no     | no    | no        |
++-------------------------------+------------+-------+---------+--------+-------+-----------+
+|:c:func:`k_thread_create`      | ?          | ?     | ?       | ?      | no    | yes       |
++-------------------------------+------------+-------+---------+--------+-------+-----------+
+|:c:func:`spi_transceive`       | yes*       | yes   | no      | no     | no    | no        |
++-------------------------------+------------+-------+---------+--------+-------+-----------+
+|:c:func:`spi_transceive_async` | yes*       | yes   | no      | no     | yes   | no        |
++-------------------------------+------------+-------+---------+--------+-------+-----------+
+
+``*`` indicates the attribute value is implicit from another attribute.
+For example a function that is not **reschedule** cannot be **sleep**;
+one that is **sleep** is necessarily **reschedule**; and one that cannot
+sleep has no use for no-wait support and is implicitly **isr-ok**.
+
+.. _api_term_reschedule:
+
+reschedule
+==========
+
+The reschedule attribute is used on a function that can reach a
+:ref:`reschedule point <scheduling_v2>` within its execution.
```

Comment:
The document linked to does not define the term "reschedule point" -- it should if that's the ref target.

What about adding something like this to the top?

```
diff --git a/doc/reference/kernel/scheduling/index.rst b/doc/reference/kernel/scheduling/index.rst
index 7a4a469ff2..b6a6c61803 100644
--- a/doc/reference/kernel/scheduling/index.rst
+++ b/doc/reference/kernel/scheduling/index.rst
@@ -9,8 +9,16 @@ to share the CPU.
 Concepts
 ********
 
-The scheduler determines which thread is allowed to execute
-at any point in time; this thread is known as the **current thread**.
+The scheduler determines which thread is allowed to execute;
+this thread is known as the **current thread**.
+
+There are various points in time when the scheduler has an opportunity to
+change the identity of the current thread. Each of these opportunities is a
+**reschedule point**. Some potential reschedule points are:
+
+- exit from ISR back into thread context
+- whenever a thread invokes :c:func:`k_yield()`
+- system timer ISR handling
 
 Whenever the scheduler changes the identity of the current thread,
 or when execution of the current thread is replaced by an ISR,
```

Reply:
This is the only comment I would consider potentially blocking. I'm deliberately not +1 or -1 at this point.